lib/blocks.php

```diff
@@ -592,3 +592,108 @@ function gutenberg_multiple_block_styles( $metadata ) {
 	return $metadata;
 }
 add_filter( 'block_type_metadata', 'gutenberg_multiple_block_styles' );
+
+if ( ! function_exists( 'wp_enqueue_block_view_script' ) ) {
+	/**
+	 * Enqueue a frontend script for a specific block.
+	 *
+	 * Scripts enqueued using this function will only get printed
+	 * when the block gets rendered on the frontend.
+	 *
+	 * @param string $block_name The block-name, including namespace.
+	 * @param array  $args       An array of arguments [handle,src,deps,ver,media].
+	 *
+	 * @return void
+	 */
+	function wp_enqueue_block_view_script( $block_name, $args ) {
+		$args = wp_parse_args(
+			$args,
+			array(
+				'handle'     => '',
+				'src'        => '',
+				'deps'       => array(),
+				'ver'        => false,
+				'in_footer'  => false,
+
+				// Additional arg to allow translations for the script's textdomain.
+				'textdomain' => '',
+			)
+		);
+
+		/**
+		 * Callback function to register and enqueue scripts.
+		 *
+		 * @param string $content When the callback is used for the render_block filter,
+		 *                        the content needs to be returned so the function parameter
+		 *                        is to ensure the content exists.
+		 * @return string Block content.
+		 */
+		$callback = static function( $content, $block ) use ( $args, $block_name ) {
+
+			// Sanity check.
+			if ( empty( $block['blockName'] ) || $block_name !== $block['blockName'] ) {
+				return $content;
+			}
+
+			// If a textdomain is defined, ensure that `wp-i18n` is enqueued.
+			if ( ! empty( $args['textdomain'] ) && ! in_array( 'wp-i18n', $args['deps'], true ) ) {
```

Comment:
We do the opposite for block scripts:
https://github.com/WordPress/wordpress-develop/blob/05141500261c845fb4dca2da28c967dc94c5e551/src/wp-includes/blocks.php#L129

The assumption was that if there is no `wp-18n` set as a dependency then it means there is no translation detected in the file so there is no need to wire translations.

Reply:
Yeah, my initial thinking here was that if they define a `textdomain`, they want translations... But the opposite is more consistent and makes sense. Changed it 👍